app/scripts/metamask-controller.js

```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
+  * Triggers decrypt, and the callback function from newUnsignedDecryptMessage.
+  *
+  * @param {Object} msgParams - The params of the message to decrypt & return to the Dapp.
+  * @returns {Promise<Object>} - A full state update.
+  */
+  decryptMessage (msgParams) {
+    log.info('MetaMaskController - decryptMessage')
+    const msgId = msgParams.metamaskId
+    // sets the status op the message to 'approved'
+    // and removes the metamaskId for decryption
+    try {
+      return this.decryptMessageManager.approveMessage(msgParams)
+        .then((cleanMsgParams) => {
+          const stripped = ethUtil.stripHexPrefix(cleanMsgParams.data)
+          const buff = Buffer.from(stripped, 'hex')
+          cleanMsgParams.data = JSON.parse(buff.toString('utf8'))
+          // decrypt the message
+          return this.keyringController.decryptMessage(cleanMsgParams)
+        })
+        .then((rawMess) => {
+        // tells the listener that the message has been decrypted
+        // and can be returned to the dapp
+          this.decryptMessageManager.setMsgStatusDecrypted(msgId, rawMess)
+          return this.getState()
+        })
+    } catch (error) {
+      log.info('MetaMaskController - eth_decryptMessage failed.', error)
+      this.decryptMessageManager.errorMessage(msgId, error)
+    }
+  }
+
+  /**
+   * Used to cancel a eth_decryptMessage type message.
+   * @param {string} msgId - The ID of the message to cancel.
+   * @param {Function} cb - The callback function called with a full state update.
+   */
+  cancelDecryptMessage (msgId, cb) {
+    const messageManager = this.decryptMessageManager
+    messageManager.rejectMsg(msgId)
+    if (cb && typeof cb === 'function') {
+      cb(null, this.getState())
+    }
+  }
+
+  // encryption_public_key methods
+
+  /**
+  * Called when a dapp uses the encryption_public_key method.
+  *
+  * @param {Object} msgParams - The params of the message to sign & return to the Dapp.
+  * @param {Function} cb - The callback function called with the signature.
+  * Passed back to the requesting Dapp.
+  */
+  async newRequestEncryptionPublicKey (msgParams, req) {
+    const promise = this.encryptionPublicKeyManager.addUnapprovedMessageAsync(msgParams, req)
+    this.sendUpdate()
+    this.opts.showUnconfirmedMessage()
+    return promise
+  }
+
+  /**
+  * Signifies a user's approval to receiving encryption public key in queue.
+  * Triggers receiving, and the callback function from newUnsignedEncryptionPublicKey.
+  *
+  * @param {Object} msgParams - The params of the message to receive & return to the Dapp.
+  * @returns {Promise<Object>} - A full state update.
+  */
+  encryptionPublicKey (msgParams) {
+    log.info('MetaMaskController - encryptionPublicKey')
+    const msgId = msgParams.metamaskId
+    // sets the status op the message to 'approved'
+    // and removes the metamaskId for decryption
+    try {
+      return this.encryptionPublicKeyManager.approveMessage(msgParams)
```

Comment:
The same two comments apply here as well - this should use `async/await`, and the catch block should re-throw the error.